ibis/pandas/udf.py

```diff
@@ -0,0 +1,192 @@
+from __future__ import absolute_import
+
+import collections
+import functools
+import itertools
+
+import six
+
+import toolz
+
+import numpy as np
+import pandas as pd
+
+from pandas.core.groupby import SeriesGroupBy
+
+# import ibis.expr.types as ir
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+
+from ibis.pandas.core import scalar_types
+from ibis.pandas.dispatch import execute_node, Dispatcher, pause_ordering
+
+
+rule_to_python_type = Dispatcher('rule_to_python_type')
+
+
+def nullable(datatype):
+    return (type(None),) if datatype.nullable is True else ()
+
+
+with pause_ordering():
+
+    @rule_to_python_type.register(dt.DataType)
+    def datatype_rule(rule, groupby=False):
+        return scalar_types
+
+    @rule_to_python_type.register(dt.Array)
+    def array_rule(rule, groupby=False):
+        return (list,)
+
+    @rule_to_python_type.register(dt.Map)
+    def map_rule(rule, groupby=False):
+        return (dict,)
+
+    @rule_to_python_type.register(dt.Struct)
+    def struct_rule(rule, groupby=False):
+        return (collections.OrderedDict,)
+
+    @rule_to_python_type.register(dt.String)
+    def string_rule(rule, groupby=False):
+        return six.string_types
+
+    @rule_to_python_type.register(dt.Integer)
+    def int_rule(rule, groupby=False):
+        return six.integer_types + (np.integer,)
+
+    @rule_to_python_type.register(dt.Floating)
+    def float_rule(rule, groupby=False):
+        return (float, np.floating)
+
+
+def signature(input_type, klass):
+    return (
+        (klass,) + rule_to_python_type(r) + nullable(r) for r in input_type
```

Comment:
I had to squint a bit at `nullable(r)` to see what it's doing, not sure if there's a way to make clearer / improve readability